frontend/src/utils/server-utils.ts

```diff
@@ -207,44 +207,45 @@ const updateLayerDatesWithValidity = (layer: ValidityLayer): DateItem[] => {
 
   const dateItemsWithValidity = momentDates.reduce(
     (acc: DateItem[], momentDate) => {
-      const endDate =
-        mode === DatesPropagation.BOTH || mode === DatesPropagation.FORWARD
-          ? momentDate.clone().add(value, 'days')
-          : momentDate.clone();
-
-      const startDate =
-        mode === DatesPropagation.BOTH || mode === DatesPropagation.BACKWARD
-          ? momentDate.clone().subtract(value, 'days')
-          : momentDate.clone();
-
-      const daysToAdd = [...Array(endDate.diff(startDate, 'days') + 1).keys()];
-
-      const days: number[] = daysToAdd
-        .map(day => startDate.clone().add(day, 'days').valueOf())
-        .filter(d => d > momentDate.valueOf());
+      let startDate = momentDate.clone();
+      let endDate = momentDate.clone();
+
+      if (mode === DatesPropagation.BOTH || mode === DatesPropagation.FORWARD) {
+        // eslint-disable-next-line fp/no-mutation
+        endDate = endDate.add(days, 'days');
+      }
+
+      if (
+        mode === DatesPropagation.BOTH ||
+        mode === DatesPropagation.BACKWARD
+      ) {
+        // eslint-disable-next-line fp/no-mutation
+        startDate = startDate.subtract(days, 'days');
+      }
+
+      const daysToAdd = Array.from(
+        { length: endDate.diff(startDate, 'days') + 1 },
+        (_, index) => startDate.clone().add(index, 'days').valueOf(),
+      );
 
-      const dateItemsToAdd: DateItem[] = days.map(dateToAdd => ({
+      const dateItemsToAdd = daysToAdd.map(dateToAdd => ({
         displayDate: dateToAdd,
         queryDate: momentDate.valueOf(),
       }));
 
       const filteredDateItems = acc.filter(
-        dateItem => days.includes(dateItem.displayDate) === false,
+        dateItem => !daysToAdd.includes(dateItem.displayDate),
       );
 
-      const mergedDateItems: DateItem[] = [
-        ...filteredDateItems,
-        ...dateItemsToAdd,
-      ];
-
-      return mergedDateItems;
+      return [...filteredDateItems, ...dateItemsToAdd];
     },
     [],
   );
 
-  const dateItems = [...dateItemsDefault, ...dateItemsWithValidity];
-
-  return sortBy(dateItems, 'displayDate');
+  return sortedUniqBy(
```

Comment:
@HarryMytilinaios @wadhwamatic why might want to think a bit harder about how we select dates here.

The underlying "tricky" question is: which data do we select if have multiple overlapping options after adding the validity periods.

- Do we select the closest real value?
- Do we select the closest past value?

Reply:
Yes I agree, but we still need to decide how to dedup them.

Reply:
@wadhwamatic we'll need your input here :)

Reply:
I think I'm following what we're trying to decide here... 

So let's say the first layer has validity range of 10-days forwards, and has 'real dates' on 1-May, 11-May, and 21-May. The latest is loaded which is valid from 21-May to 31-May. 

Then a second layer is loaded which is valid on 11-May and 24-May with three days backwards. Even though 11-May is an 'exact' match for the reference date, we should use 24-May. 

Basically any valid date, whether it's the 'real date' or an interpreted date should be treated the same. We're really trying to show that the date of observation is not necessarily the only correct date for a measurement.

I hope I understood this right and what I'm saying is clear!

Reply:
I don't really follow here what are you saying I am gonna give you an example. We have 2 dates  for 2 different layers `11-May` and `21-May` and we have `forward` 20 days at the first and 10 days `both` at the second. The funcionality above takes every single default date in our case `11-May` and `21-May` and tries and adds or subtracts or both the number of days each specific date has a validity. After those calculations in our particular case we would end up with an array with duplicated dates firstly from the `forward` of the first date and secondly of the `both` of the second date. Those dates will be duplicates, the above code takes a unique of those. What exactly do we care if we will keep the layer's date overlap or second's? Am I missing something here?

Reply:
I am sorry but I don't follow here. Every layer has its own date array. The duplicates were being on the same layer's date array. We can have duplicate entries in the date array of different layers. This is how the intersected dates occur. We cannot have duplicate dates on the same layer's date array. Is there something I am missing here?

Reply:
So @ericboucher explained to me what we are reffering here, and right now this is  exactly what this functionality does. The `sortUniqBy` after the array of dates is sorted takes the most recent actual date. so all good. I didn't understand at the first place what you guys were reffering.